### PR TITLE
Allows rebin of masks

### PR DESCRIPTION
**Pull request by dtasev**
- Adds the type tuple needed in core/apply.h
- Add test for 1d/2d mask rebinning, and with bin weights
- Adds Dataset 1D rebin test
- Rebin only masks with the same dimension - keep unrelated masks
- Adds DataArray mask setter

### CODE REVIEW

File: core/rebin.cpp [hunk 1]
```diff
@@ -123,7 +126,7 @@ Variable rebin(const VariableConstProxy &var, const Dim dim,
   // Rebin could also implemented for count-densities. However, it may be better
   // to avoid this since it increases complexity. Instead, densities could
   // always be computed on-the-fly for visualization, if required.
-  expect::unit(var, units::counts);
+  // expect::unit(var, units::counts);
```
Comment (dtasev) [inline core/rebin.cpp]: @SimonHeybrock Not sure what to do with this check, except to remove it. Any suggestions?

File: core/dataset_operations.cpp [hunk 2]
```diff
@@ -114,6 +114,11 @@ DataArray rebin(const DataConstProxy &a, const Dim dim,
                 const VariableConstProxy &coord) {
   auto rebinned = apply_to_data_and_drop_dim(
       a, [](auto &&... _) { return rebin(_...); }, dim, a.coords()[dim], coord);
+
+  for (auto &&[name, mask] : a.masks()) {
+    rebinned.setMask(name, rebin(mask, dim, a.coords()[dim], coord));
```
Comment (SimonHeybrock) [inline core/dataset_operations.cpp]: Shouldn't this be done only if `mask` depends on `dim`?

File: core/include/scipp/core/dataset.h [hunk 3]
```diff
@@ -1100,6 +1100,13 @@ class SCIPP_CORE_EXPORT DataArray {
     setCoord(dim, Variable(coord));
   }
 
+  void setMask(const std::string &name, Variable mask) {
+    m_holder.setMask(name, std::move(mask));
+  }
+  void setMask(const std::string &name, const VariableConstProxy &mask) {
+    setMask(name, Variable(mask));
+  }
+
```
Comment (SimonHeybrock) [inline core/include/scipp/core/dataset.h]: I think I had avoided adding these to `DataArray`, also for labels, attrs, etc. I recommend to use `array.masks().set(name, mask)` instead. That is also more in-line with how we write Python code.

File: core/rebin.cpp [hunk 4]
```diff
@@ -6,6 +6,8 @@
 #include "scipp/core/except.h"
 #include "scipp/core/variable.h"
 
+using mask_rebinning_t = std::tuple<bool, bool, double, double>;
+
```
Comment (SimonHeybrock) [inline core/rebin.cpp]: Put this closer to where it is used?

File: core/rebin.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply core/rebin.cpp]: Maybe it makes sense to allow dimensionless in addition to counts, but forbid anything else?

File: core/test/rebin_test.cpp [hunk 5]
```diff
@@ -96,3 +96,57 @@ TEST_F(RebinTest, keeps_unrelated_labels_but_drops_others) {
...
+                                        false, false, false, false});
+};
+
+TEST_F(RebinMask1DTest, mask_1d) {
+
+  const auto edges = makeVariable<double>({Dim::X, 5}, {1, 3, 5, 7, 10});
+  const auto expected =
+      makeVariable<bool>({Dim::X, 4}, {false, true, false, false});
+
+  const auto result = rebin(mask, Dim::X, x, edges);
```
Comment (SimonHeybrock) [inline core/test/rebin_test.cpp]: Hmmm, shouldn't this fail in a size check? The mask has the same length as the old `x` edges, but it should be 1 shorter? You have not changed that code, so it might be an existing bug?

File: core/test/dataset_operations_test.cpp [hunk 6]
```diff
@@ -105,3 +105,24 @@ TEST(DatasetOperationsTest, mean_three_dims) {
...
+  ds.setCoord(Dim::X, makeVariable<double>({Dim::X, 5}, {1, 2, 3, 4, 5}));
+  ds.setData("data_x", makeVariable<double>({Dim::X, 5}, {1, 2, 3, 4, 5}));
+
+  ds.setMask("mask_x", makeVariable<bool>({Dim::X, 5},
+                                          {false, false, true, false, false}));
+  ds.setMask("mask_y", makeVariable<bool>({Dim::Y, 5},
+                                          {false, false, true, false, false}));
+
+  const auto edges = makeVariable<double>({Dim::X, 3}, {1, 3, 5});
+  const Dataset result = rebin(ds, Dim::X, edges);
```
Comment (SimonHeybrock) [inline core/test/dataset_operations_test.cpp]: Same here, this should have failed, since `Dim::X` is not actually bin edges?

File: core/test/rebin_test.cpp (hunk 5 above)
Comment (dtasev) [reply core/test/rebin_test.cpp]: I did see the fail only when writing the 2D test, that's why the size of Dim::X is 6 there. I don't think the check is properly working for 1D, but it's the same code isn't it? I wondered myself why [this](https://github.com/scipp/scipp/blob/master/core/rebin.cpp#L136-L138) check wasn't failing for not bin-edges:

File: common/include/scipp/common/except.h [hunk 7]
```diff
@@ -5,8 +5,12 @@
 #ifndef SCIPP_COMMON_EXCEPT_H
 #define SCIPP_COMMON_EXCEPT_H
 
+#include <initializer_list>
+#include <sstream>
```
Comment (SimonHeybrock) [inline common/include/scipp/common/except.h]: Why is this required? I think it is quite a large header that we should avoid including everywhere.

File: common/include/scipp/common/string.h [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+template <class T> std::string to_string(const std::initializer_list<T> items) {
+  std::stringstream ss;
+
+  for (const auto &item : items) {
+    ss << to_string(item) << " ";
+  }
+
+  return ss.str();
+}
```
Comment (SimonHeybrock) [inline common/include/scipp/common/string.h]: One way to avoid the `sstream` include would be to put the template definition into the `cpp` and use explicit template instantiation for the (handful?) of required types.

File: core/include/scipp/core/except.h [hunk 9]
```diff
@@ -139,6 +149,11 @@ template <class T> void unit(const T &object, const units::Unit &unit) {
   expect::equals(object.unit(), unit);
 }
 
+template <class T>
+void unit_one_of(const T &object, std::initializer_list<units::Unit> units) {
```
Comment (SimonHeybrock) [inline core/include/scipp/core/except.h]: Name these `.._any_of`, for consistency with `std::any_of`, which does a similar thing??

File: core/include/scipp/core/except.h [hunk 10]
```diff
@@ -108,6 +109,15 @@ template <class A, class B> void equals(const A &a, const B &b) {
     throw scipp::except::MismatchError(a, b);
 }
 
+template <class A, class B>
+void equals_one_of(const A &a, const std::initializer_list<B> possible) {
+  const auto result = std::find(possible.begin(), possible.end(), a);
+
+  // if none were equal throw the exception
+  if (result == possible.end())
```
Comment (SimonHeybrock) [inline core/include/scipp/core/except.h]: ```if (std::find(possible.begin(), possible.end(), a) == possible.end())```
saves some lines and a named variable?
